lib/connectors/memory.js

```diff
@@ -250,6 +250,46 @@ Memory.prototype.updateOrCreate = function(model, data, options, callback) {
     }
   });
 };
+Memory.prototype.patchOrCreateWithWhere =
+  Memory.prototype.upsertWithWhere = function(model, where, data, options, callback) {
+    var self = this;
+    var element = Object.keys(where)[0];
+    console.log (element);
+    var filter = { where: where };
+    console.log(filter);
+    console.log(model);
+    if (!data.hasOwnProperty(element)) {
+      data[element] = where[element];
+    }
+    var primaryKey = this.idName(model);
+    self.all(model, filter, options, function(err, instances) {
+      if (err) callback(err);
+      else {
+        var modelsLength = instances.length;
+        if (modelsLength === 0 && data.hasOwnProperty(primaryKey)) {
+          self.create(model, data, options, function(err, id) {
+            self.setIdValue(model, data, id);
+            callback(err, data, { isNewInstance: true });
+          });
+        } else if (modelsLength == 0 && data.hasOwnProperty(primaryKey) == false) {
+          var error = new Error('Error message: No matching instance found.Found error ' +
+            'while trying to create a new instance. Primary key is not provided ' +
+            'in the payload');
```

Comment:
I have a different opinion. Error messages should be helpful in the first place. It's much better to have a long message that makes it easy to understand what's the problem, compared to a short message that lefts the reader in dark.

A great error message also includes instructions on how to fix the problem.